src/api/environment.cc

```diff
@@ -330,33 +334,149 @@ Environment* CreateEnvironment(IsolateData* isolate_data,
   // options than the global parse call.
-  std::vector<std::string> args(argv, argv + argc);
-  std::vector<std::string> exec_args(exec_argv, exec_argv + exec_argc);
-  // TODO(addaleax): Provide more sensible flags, in an embedder-accessible way.
   Environment* env = new Environment(
       isolate_data,
       context,
       args,
       exec_args,
-      static_cast<Environment::Flags>(Environment::kIsMainThread |
-                                      Environment::kOwnsProcessState |
-                                      Environment::kOwnsInspector));
-  env->InitializeLibuv(per_process::v8_is_profiling);
-  if (env->RunBootstrapping().IsEmpty())
+      flags,
+      thread_id);
+  if (flags & EnvironmentFlags::kOwnsProcessState) {
+    env->set_abort_on_uncaught_exception(false);
+  }
+
+  if (env->RunBootstrapping().IsEmpty()) {
+    FreeEnvironment(env);
     return nullptr;
+  }
+
   return env;
 }
 
 void FreeEnvironment(Environment* env) {
-  env->RunCleanup();
+  {
+    // TODO(addaleax): This should maybe rather be in a SealHandleScope.
+    HandleScope handle_scope(env->isolate());
+    Context::Scope context_scope(env->context());
+    env->set_stopping(true);
+    env->stop_sub_worker_contexts();
+    env->RunCleanup();
+    RunAtExit(env);
+  }
+
+  // This call needs to be made while the `Environment` is still alive
+  // because we assume that it is available for async tracking in the
+  // NodePlatform implementation.
+  MultiIsolatePlatform* platform = env->isolate_data()->platform();
+  if (platform != nullptr)
+    platform->DrainTasks(env->isolate());
+
   delete env;
 }
 
+InspectorParentHandle::~InspectorParentHandle() {}
+
+// Hide the internal handle class from the public API.
+#if HAVE_INSPECTOR
+struct InspectorParentHandleImpl : public InspectorParentHandle {
+  std::unique_ptr<inspector::ParentInspectorHandle> impl;
+
+  explicit InspectorParentHandleImpl(
+      std::unique_ptr<inspector::ParentInspectorHandle>&& impl)
+    : impl(std::move(impl)) {}
+};
+#endif
+
+NODE_EXTERN std::unique_ptr<InspectorParentHandle> GetInspectorParentHandle(
+    Environment* env,
+    ThreadId thread_id,
+    const char* url) {
+  CHECK_NOT_NULL(env);
+  CHECK_NE(thread_id.id, static_cast<uint64_t>(-1));
+#if HAVE_INSPECTOR
+  return std::make_unique<InspectorParentHandleImpl>(
+      env->inspector_agent()->GetParentHandle(thread_id.id, url));
+#else
+  return {};
+#endif
+}
+
+void LoadEnvironment(Environment* env) {
+  USE(LoadEnvironment(env,
+                      StartExecutionCallback{},
+                      {}));
+}
+
+MaybeLocal<Value> LoadEnvironment(
+    Environment* env,
+    StartExecutionCallback cb,
+    std::unique_ptr<InspectorParentHandle> inspector_parent_handle) {
+  env->InitializeLibuv(per_process::v8_is_profiling);
+  env->InitializeDiagnostics();
+
+#if HAVE_INSPECTOR
+  if (inspector_parent_handle) {
+    env->InitializeInspector(
+        std::move(static_cast<InspectorParentHandleImpl*>(
```

Comment:
I just remember why this needs to be done before bootstrapping - otherwise `--inspect-brk-node` breaks

Reply:
@joyeecheung Hm yeah … it feels like there should be tests for that? Otherwise this is just going to be broken from time to time one way or another…

Reply:
@addaleax I guess I was trying to remind about that in my TODO removed in this PR (in a way that's quite difficult to remember once out of context) because I've ran into this before when tweaking with my snapshot integration branch (this should be started before bootstrapping in the snapshot creator, and after bootstrapping in deserialization mode at run time). Not exactly sure how this can be tested though (checking the stack frame of the first paused statement?), or I think there were some discussions about not adding a test since it was a developer feature?

Reply:
I’ve opened https://github.com/nodejs/node/pull/32672 to address this … with a test, because otherwise this ends up being broken again at some point